main/src/library/Fixpoint3/Phase/Hoisting.flix

```diff
@@ -0,0 +1,642 @@
+                        // The 'output' variables from `Functional` may not have been bound earlier.
+                        // Forcefull removal from `termMap` is therefore safe.
+                        Vector.range(0, arity) |>
+                        Vector.forEach(i -> {
+                            match MutDisjointSets.find((rowVar, i), equalitySets) {
+                                case Some(rep) => MutMap.remove(rep, termMap)
+                                case None => bug!("In Fixpoint.Phase.Hoisting: Everything should be in the equalitySet")
+                            }
+                        });
+                        let hoistedToHere = List.filter(x -> not isBoolGround(termMap, equalitySets, x), relevant);
+
+                        if (List.nonEmpty(hoistedToHere)) {
+                            Some(RelOp.Functional(rowVar, func, terms, RelOp.If(List.toVector(hoistedToHere) |> Vector.map(replaceByRepBoolExp(termMapCopy, equalitySets)), v), arity))
+                        } else {
+                            Some(RelOp.Functional(rowVar, func, terms, v, arity))
+                        }
+                    }
+                }
+            case RelOp.Project(_, _) => Some(op)
+            case RelOp.If(boolExps, body) =>
+                // Just note all `boolExps`. They will be placed as high as possible in the AST, by other cases.
+                // If `boolExps` contains no variables push it all the way to the top, otherwise register it for
+                // all variables occuring in `boolExp`.
+                boolExps |>
+                Vector.forEach(b -> match b {
+                    // Don't collect Eq's. They are handled separately to ensure we refer to the earliest evaluated expression.
+                    case BoolExp.Eq(_, _) => ()
+                    case _ =>
+                        let vars = collectBoolVariables(termMap, equalitySets, b);
+                        if (Set.size(vars) == 0)
+                            MutList.push(b, independentBool)
+                        else
+                            vars |> Set.forEach(v -> {
+                                let boolList = MutMap.getOrElsePut(v, MutList.empty(rc), idToBool);
+                                MutList.push(b, boolList)})
+                });
+                hoistOpRec(body)
+        }
+
+    ///
+    /// Returns the set of `RowVar` in `bool` which are mapped to other `RowVar` with respect to `termMap`.
+    ///
+    /// Ignores `Eq`, as `Eq`'s are seen as a special case and will always be a comparison between loads and
+    /// loads/constants.
+    ///
+    /// As an example `Guard2(x[0], y[3])` will return `{x}` if `termMap` is e.g. `[y[3] => 8]`.
+    /// If `termMap` contained `y[3]=> z[3]` then `{x, z}` will be returned.
+    ///
+    /// The purpose is to return the `RowVar`'s in the AST which are possible places `bool` can be hoisted to.
+    ///
+    def collectBoolVariables(
```

Comment:
I don't know how hot-path this function is, but It'll probably be faster in tail-recursive form (with a work list and accumulator)